### PR TITLE
Added API call timeout

### PR DESCRIPTION
**Pull request by marcellobarile**
* Added a new setting to set the CURL timeout (useful for #346)
* Added a new setting to enable the "webinar" mode by default when creating a new activity

### CODE REVIEW

File: lang/en/zoom.php [hunk 1]
```diff
@@ -44,6 +44,8 @@
 $string['apiidentifier_desc'] = 'The identifier field to use when making a call to the Zoom API';
 $string['apikey'] = 'Zoom API key';
 $string['apikey_desc'] = '';
+$string['apitimeout'] = 'Zoom API timeout';
+$string['apitimeout_desc'] = 'Amount of seconds before to consider an API call expired. Default is 60.';
```
Comment (jrchamp) [inline lang/en/zoom.php:48]:
```suggestion
$string['apitimeout_desc'] = 'Number of seconds to wait before closing the API connection. Default is 60.';
```

File: mod_form.php [hunk 2]
```diff
@@ -301,7 +301,7 @@ public function definition() {
                     }
                     $mform->addElement('advcheckbox', 'webinar', get_string('webinar', 'zoom'),
                             get_string('webinarthisis', 'zoom'), $webinarattr);
-                    $mform->setDefault('webinar', 0);
+                    $mform->setDefault('webinar', $config->webinardefault );
```
Comment (jrchamp) [inline mod_form.php:304]: Extra space
```suggestion
                    $mform->setDefault('webinar', $config->webinardefault);
```

File: settings.php [hunk 3]
```diff
@@ -68,6 +68,10 @@
             get_string('apisecret_desc', 'mod_zoom'), '');
     $settings->add($apisecret);
 
+    $apitimeout = new admin_setting_configtext('zoom/apitimeout', get_string('apitimeout', 'mod_zoom'),
+            get_string('apitimeout_desc', 'mod_zoom'), '');
```
Comment (jrchamp) [inline settings.php:72]: Seems like this should be a number instead of text and default to 60.

**Review by jrchamp — COMMENTED** [review]
Were you able to test and confirm that you were getting timeouts? I have some comments, but overall the code makes sense to me.

File: classes/webservice.php [hunk 4]
```diff
@@ -216,6 +222,14 @@ private function make_call($path, $data = array(), $method = 'get') {
         $token = \Firebase\JWT\JWT::encode($payload, $this->apisecret, 'HS256');
         $curl->setHeader('Authorization: Bearer ' . $token);
 
+        $calltimeout = get_config('zoom', 'apitimeout');
+        if (empty($calltimeout)) {
+          $calltimeout = self::CALL_TIMEOUT;
+        }
+
+        $curl->setopt(array('CURLOPT_CONNECTTIMEOUT' => $calltimeout));
+        $curl->setopt(array('CURLOPT_TIMEOUT' => $calltimeout));
```
[the remark this answers is not in the document]
Comment (marcellobarile) [reply classes/webservice.php]: fixed, silly me :)

File: lang/en/zoom.php (hunk 1 above)
Comment (marcellobarile) [reply lang/en/zoom.php]: updated, thanks for the hint

File: settings.php (hunk 3 above)
Comment (marcellobarile) [reply settings.php]: yup, done

**Comment by marcellobarile** [comment]
> Were you able to test and confirm that you were getting timeouts? I have some comments, but overall the code makes sense to me.

Honestly, I did an empirical test; since I bumped the timeout to 60 seconds the meetings are not marked as expired anymore.
I thought that because the change was not destructive, I could proceed with it in any case.
I know it's suboptimal so I would understand if before proceeding with the merge you would ask for something objective :)

**Comment by jrchamp** [comment]
I'll try and rebase this to avoid the grunt issues that were fixed in the lasted changes.

**Comment by jrchamp** [comment]
I'll try and rebase this to avoid the grunt issues that were fixed in the lasted changes.

**Comment by jrchamp** [comment]
Based on the testing that I've done, cURL doesn't have a low connected timeout (the amount of time after a successful connection is established before it gives up). The default seems to be `0` which means "no limit". The default connection timeout (the amount of time it waits for a connection to be established) seems to be somewhere about 250 seconds. It looks like Moodle's `curl` class that we are using is initialized with a connection timeout of 30 seconds: https://github.com/moodle/moodle/blob/ad6dc71c56280c3aad236eb7b0f6650e00eecf34/lib/filelib.php#L3233

30 seconds should be enough time to establish a connection (most of the places where it gets overridden in Moodle set the value even lower to somewhere between 2 and 10 seconds). The changes suggested here would - by default - increase the connection timeout from 30 seconds to 60 seconds, and reduce the overall timeout from unlimited to 60 seconds.

As an interesting side note: PHP's `default_socket_timeout` ini value (which does default to 60 seconds) is not used by cURL in any of my testing.

**Comment by jrchamp** [comment]
I've opened #367 to split out the functionality that is ready to be merged (the webinar by default option). As the core of this PR is focused on timeouts, it seemed better to leave that functionality here.

**Comment by sgrandh3** [comment]
I did regression testing and nothing seems to break, everything looks good.

**Comment by jrchamp** [comment]
Thank you @marcellobarile for your work on this issue. While the current code doesn't seem to break anything, it also doesn't appear to provide direct benefit towards resolving the the original issue. At this time, let's close the issue until evidence exists.

**Pull request closed without merging**